tools/cosmovisor/cmd/cosmovisor/prepare_upgrade.go

```diff
@@ -0,0 +1,66 @@
+package main
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/spf13/cobra"
+
+	"cosmossdk.io/tools/cosmovisor"
+	"cosmossdk.io/x/upgrade/plan"
+)
+
+func NewPrepareUpgradeCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "prepare-upgrade",
+		Short: "Prepare for the next upgrade",
+		Long: `Prepare for the next upgrade by downloading and verifying the upgrade binary.
+This command will download the binary specified in the upgrade-info.json file,
+verify its checksum, and place it in the appropriate directory for Cosmovisor to use.`,
+		RunE:         prepareUpgradeHandler,
+		SilenceUsage: false,
+		Args:         cobra.NoArgs,
+	}
+	return cmd
+}
+
+func prepareUpgradeHandler(cmd *cobra.Command, _ []string) error {
+	configPath, err := cmd.Flags().GetString(cosmovisor.FlagCosmovisorConfig)
+	if err != nil {
+		return fmt.Errorf("failed to get config flag: %w", err)
+	}
+
+	cfg, err := cosmovisor.GetConfigFromFile(configPath)
+	if err != nil {
+		return fmt.Errorf("failed to get config: %w", err)
+	}
+
+	logger := cfg.Logger(cmd.OutOrStdout())
+	upgradeInfo, err := cfg.UpgradeInfo()
```

Comment:
That won't always work. 
Cosmovisor creates manually the upgrade info when you use `add upgrade` for the specific height (which is cosmovisor specific handling).
Normally, the upgrading info is only saved just before chain halt (via governance upgrade proposal for instance): https://github.com/cosmos/cosmos-sdk/blob/787ee69/x/upgrade/keeper/abci.go#L76-L78

Reply:
Ok, just one doubt, can it happen that the url of the binary is not present in the blockchain query?